File: unix/boot/spp/rpp/ratlibf/enter.f
      subroutine enter (symbol, info, st)
      integer symbol (100)
      integer info (100)
      integer st
      integer mem( 60000)
      common/cdsmem/mem
      integer i, nodsiz, j
      integer stlu, length
      integer node, pred
      integer dsget
      nodsiz = mem (st)
      if (.not.(stlu (symbol, node, pred, st) .eq. 0))goto 23000
      node = dsget (1 + nodsiz + length (symbol) + 1)
      mem (node + 0) = 0
      mem (pred + 0) = node
      i = 1
      j = node + 1 + nodsiz
23002 if (.not.(symbol (i) .ne. -2))goto 23003
      mem (j) = symbol (i)
      i = i + 1
      j = j + 1
      goto 23002
23003 continue
      mem (j) = -2
23000 continue
      i = 1
23004 if (.not.(i .le. nodsiz))goto 23006
      j = node + 1 + i - 1
      mem (j) = info (i)
23005 i = i + 1
      goto 23004
23006 continue
      return
      end

File: unix/boot/spp/rpp/ratlibr/enter.r
include	defs

# enter --- place a symbol in the symbol table, updating if already present

   subroutine enter (symbol, info, st)
   character symbol (ARB)
   integer info (ARB)
   pointer st

   DS_DECL(Mem, MEMSIZE)

   integer i, nodsiz, j
   integer stlu, length

   pointer node, pred
   pointer dsget

   nodsiz = Mem (st)

   if (stlu (symbol, node, pred, st) == NO) {
      node = dsget (1 + nodsiz + length (symbol) + 1)
      Mem (node + ST_LINK) = LAMBDA
      Mem (pred + ST_LINK) = node
      i = 1
      j = node + ST_DATA + nodsiz
      while (symbol (i) != EOS) {
	 Mem (j) = symbol (i)
	 i = i + 1
	 j = j + 1
	 }
      Mem (j) = EOS
      }

   for (i = 1; i <= nodsiz; i = i + 1) {
      j = node + ST_DATA + i - 1
      Mem (j) = info (i)
      }

   return
   end
